### PR TITLE
Berk master for complete working model. 

### PR DESCRIPTION
**Pull request by 1ozturkbe**
@bqpd we should merge this into master as soon as we can so we have something that makes sense to show to NASA folks a few weeks from now. I hope the merge conflicts will not be a massive headache.

### CODE REVIEW

**Comment by 1ozturkbe** [comment]
Oooh nice, you already merged. Are you getting 21W as a result?

**Comment by bqpd** [comment]
Merged. Biggest change was moving the substitutions and materials back into the model declarations, because `server.py` just creates a vanilla Layer right now, though that could be changed. Unfortunately the geometry this generates does not build with `demo_tile.csm`, and so this can't be merged to master yet!

**Comment by 1ozturkbe** [comment]
Well let's figure out how to generate demo_tile from this. What needs to be done?

**Comment by bqpd** [comment]
I'm checking the dumb stuff first, like that I have the right substitutions...

**Comment by bqpd** [comment]
You got 21W on the (4,4) case that's in designHX.py, right?

**Comment by 1ozturkbe** [comment]
I believe so.

**Comment by bqpd** [comment]
What do you get on (3,3)?

**Comment by 1ozturkbe** [comment]
Didn't realize you were running 3x3. Will check.

**Comment by 1ozturkbe** [comment]
Wait, you realize that you merged master into my branch, and overwrote a bunch of my changes.

**Comment by bqpd** [comment]
...that's how github pull requests work, sorry! We can create a new branch for you to work on (or copy this branch to a new name if you wanted to continue working on berkMaster)

**Comment by 1ozturkbe** [comment]
Alright, well, I went back up one commit, and I got 18.5 W for the 3x3.

**Comment by bqpd** [comment]
Oh just noticed you're using drag in the objective function. I think that's pretty unreasonable...it's pretty clear that we want to maximize heat transfer, not heat transfer per total drag, I think.

**Comment by bqpd** [comment]
I can try solving with that objective function to get comparable results, but I don't think it provides a good interface for users.

**Comment by bqpd** [comment]
The only modeling change I saw was the addition of a porosity variable...

**Comment by 1ozturkbe** [comment]
Ok, but how do you determine what the right 'drag' is? That's a whole other ball game, and it's not at all clear what the value should be.

**Comment by bqpd** [comment]
That's true, but I think that's a reason to give the choice to the user and set a reasonable default.

**Comment by 1ozturkbe** [comment]
What we have in this objective is a kind of efficiency parameter which I think makes more sense without a context for this HX to operate in.

**Comment by bqpd** [comment]
I'm down! How should we do objective weighting? Right now the ratio is heavily skewed to minimizing drag, since it's possible to get that much closer to 0 / logspace -inf

**Comment by bqpd** [comment]
update: yeah, max heat transfer wants a porosity of 0.8. I'll update that constant and try building again...

**Comment by bqpd** [comment]
ugh, wait, nvm, still not tiling.

**Comment by bqpd** [comment]
I think it might be n_fins; it's free on master, but set to 5 on this branch

**Comment by 1ozturkbe** [comment]
Does it collapse to 1, or is it an interior solution?

**Comment by bqpd** [comment]
checking!

**Comment by bqpd** [comment]
re: drags, I set them to a hundredth of a Newton and I get hundreds of watts of heat transfer...when they're in the objective they go to like micronewtons!

**Comment by bqpd** [comment]
It's an interior solution, n_fins of 3.746

**Comment by 1ozturkbe** [comment]
Haha yeah, what I'd expect. I bet it also makes the solidity go down.

**Comment by 1ozturkbe** [comment]
ALso, heck yes for 3.76!! Interiors are good!

**Comment by 1ozturkbe** [comment]
Looks really good to me. Once we figure out the tiling I think it's good to merge. Lmk if there is anything you need on my part.

**Comment by bqpd** [comment]
Yup, freeing n_fins gets it tiling again. I'll play around with it to confirm, then push and merge to master...do you want to start working from master again and create PRs when you have new models? (I guess I'm kinda the integration test, until you get it running on a machine there...)

**Comment by bqpd** [comment]
Of note: 0.01 N drags in fact give almost 600W of heat transfer...

**Comment by 1ozturkbe** [comment]
Great to hear. I prefer PRs anyways, good idea. Also, I'm pretty sure Bob has the tiling working on his machine now... he tried to demo it today on my branch but obviously the serve_csm stuff had not been merged. I can perhaps try tomorrow. 
And 600W sounds right, 20m/s of flow is really fast.

**Comment by bqpd** [comment]
Phew! Oh one other thing while I have you here :p I tried changing the 2D interp to be grid-based (since we have a grid) instead of linear splines on a triangulation, but it stopped tiling :( I'm really not sure why, the tiling is damnably finicky, but if it stops being so & we can get away with that I think that might be a better interpolation to use.

File: heatexchanger/rectpipe.py [hunk 1]
```diff
@@ -111,7 +111,7 @@ def setup(self, Nsegments, Nfins, fluid, increasingT):
                 (Pf/Pf_ref)**0.155 >= (0.475*(Re[-1]/Re_ref)**0.00121
                                        + 0.0338*(Re[-1]/Re_ref)**-0.336),
 
-                D >= fr * Nfins * A_seg[0]
+                D >= fr * Nfins * A[0]
```
Comment (bqpd) [inline heatexchanger/rectpipe.py]: @1ozturkbe Oh, and I can't get it to tile with this as `A[0]`, which is maddening, so master will continue to use `A_seg[0]` :(

**Comment by 1ozturkbe** [comment]
Ugh, that merge was so bad, it overwrote so many different things that I had done. I am restoring my original branch to make sure that my updates are merged correctly.

**Comment by 1ozturkbe** [comment]
Can you please be more careful in the future @bqpd? It was actually a really bad merge. Master wasn't even running.

**Comment by 1ozturkbe** [comment]
I have gotten a working master now, and will get my original changes back at some point. But for now this will do.

**Comment by bqpd** [comment]
Sorry, I thought you wanted those changes onto master ASAP! Are you working off the current master now?

**Comment by 1ozturkbe** [comment]
It's ok. I'm on master now, and it runs. I just don't want to make your serve_csm stuff break on master. Is there something I should be particularly careful about?

**Comment by bqpd** [comment]
I don't think so! Just make sure to add variables you want users to see to the list in `layer.py`...come to think of it, I should have added the max_porosity to that...

**Pull request merged**